crates/vm/levm/src/memory.rs

```diff
@@ -1,198 +1,290 @@
+        }
     }
 
-    let at_offset: usize = at_offset
-        .try_into()
-        .map_err(|_err| ExceptionalHalt::VeryLargeNumber)?;
-
-    for (byte_to_store, memory_slot) in data.iter().zip(
-        memory
-            .get_mut(
-                at_offset
-                    ..at_offset
-                        .checked_add(data_size)
-                        .ok_or(InternalError::Overflow)?,
-            )
-            .ok_or(OutOfBounds)?
-            .iter_mut(),
-    ) {
-        *memory_slot = *byte_to_store;
-    }
-    Ok(())
-}
+    /// Load N bytes from the given offset.
+    #[inline(always)]
+    pub fn load_range_const<const N: usize>(&mut self, offset: usize) -> Result<[u8; N], VMError> {
+        let new_size = offset.checked_add(N).ok_or(OutOfBounds)?;
+        self.resize(new_size)?;
 
-pub fn try_copy_within(
-    memory: &mut Memory,
-    from_offset: U256,
-    to_offset: U256,
-    size: usize,
-) -> Result<(), VMError> {
-    if size == 0 {
-        return Ok(());
+        let true_offset = offset.checked_add(self.current_base).ok_or(OutOfBounds)?;
+
+        let buf = self.buffer.borrow();
+        // SAFETY: resize already makes sure bounds are correct.
+        #[allow(unsafe_code)]
+        unsafe {
+            Ok(*buf
+                .get_unchecked(true_offset..(true_offset.wrapping_add(N)))
+                .as_ptr()
+                .cast::<[u8; N]>())
+        }
     }
 
-    let from_offset: usize = from_offset
-        .try_into()
-        .map_err(|_err| ExceptionalHalt::VeryLargeNumber)?;
-    let to_offset: usize = to_offset
-        .try_into()
-        .map_err(|_err| ExceptionalHalt::VeryLargeNumber)?;
-    try_resize(
-        memory,
-        to_offset
-            .max(from_offset)
-            .checked_add(size)
-            .ok_or(InternalError::Overflow)?,
-    )?;
+    /// Load a word from at the given offset.
+    #[inline(always)]
+    pub fn load_word(&mut self, offset: usize) -> Result<U256, VMError> {
+        let value: [u8; 32] = self.load_range_const(offset)?;
+        Ok(u256_from_big_endian_const(value))
+    }
 
-    let mut temporary_buffer = vec![0u8; size];
-    for i in 0..size {
-        if let Some(temporary_buffer_byte) = temporary_buffer.get_mut(i) {
-            *temporary_buffer_byte = *memory
-                .get(from_offset.checked_add(i).ok_or(InternalError::Overflow)?)
-                .unwrap_or(&0u8);
+    /// Stores the given data and data size at the given offset.
+    ///
+    /// Internal use.
+    #[inline(always)]
+    fn store(&self, data: &[u8], at_offset: usize, data_size: usize) -> Result<(), VMError> {
+        if data_size == 0 {
+            return Ok(());
         }
+
+        let real_offset = self.current_base.wrapping_add(at_offset);
+
+        let mut buffer = self.buffer.borrow_mut();
+
+        let real_data_size = data_size.min(data.len());
+
+        // SAFETY: Used internally, resize always called before this function.
```

Comment:
If this is the case maybe we should add a resize_and_store function that does this and ensures we won't mess up in the future?

Or even bring the resize inside if there's no notion of store without resize.

Reply:
resize is also used externally by one opcode, personally i think it's not worth it to do this, store is not pub and there are only 3 uses which can be checked quickly, furthermore the memory code shouldn't change much bar the need of another refactor.